s3upload.js

```diff
@@ -32,16 +37,19 @@ function S3Upload(options) {
             this[option] = options[option];
         }
     }
-    var files = this.fileElement ? this.fileElement.files : this.files || [];
-    this.handleFileSelect(files);
+    this.handleFileSelect(this.files);
 }
 
 S3Upload.prototype.handleFileSelect = function(files) {
     var result = [];
+    var that = this;
+    var next = function (file) {
+        that.onProgress(0, 'Waiting', file);
+        result.push(that.uploadFile(file));
+    }
     for (var i=0; i < files.length; i++) {
         var file = files[i];
-        this.onProgress(0, 'Waiting', file);
-        result.push(this.uploadFile(file));
+        this.onBeforeUpload(file, next)
```

Comment:
Oh I see.  The name `onBeforeUpload` feels a little misleading then, since it sounds more like an event listener, and event listeners can't typically change what the firing code is doing.  How about calling it `preprocess` instead of `onBeforeUpload`?